query/src/sql/optimizer/heuristic/prune_columns.rs

```diff
@@ -0,0 +1,178 @@
+            ))
+        }
+        RelOperator::EvalScalar(p) => {
+            let mut used = vec![];
+            // Only keep columns needed by parent plan.
+            p.items.iter().for_each(|s| {
+                if !required.contains(&s.index) {
+                    return;
+                }
+                used.push(s.clone());
+                s.scalar.used_columns().iter().for_each(|c| {
+                    required.insert(*c);
+                })
+            });
+            Ok(SExpr::create_unary(
+                RelOperator::EvalScalar(EvalScalar { items: used }),
+                keep_required_columns(expr.child(0)?, required)?,
+            ))
+        }
+        RelOperator::Filter(p) => {
+            let used = p.predicates.iter().fold(required, |acc, v| {
+                acc.union(&v.used_columns()).cloned().collect()
+            });
+            Ok(SExpr::create_unary(
+                RelOperator::Filter(p.clone()),
+                keep_required_columns(expr.child(0)?, used)?,
+            ))
+        }
+        RelOperator::Aggregate(p) => {
+            let mut used = vec![];
+            for item in &p.aggregate_functions {
+                if required.contains(&item.index) {
+                    for c in item.scalar.used_columns() {
+                        required.insert(c);
+                    }
+                    used.push(item.clone());
+                }
+            }
+            p.group_items.iter().for_each(|i| {
+                // If the group item comes from a complex expression, we only include the final
+                // column index here. The used columns will be included in its EvalScalar child.
+                required.insert(i.index);
+            });
+            Ok(SExpr::create_unary(
+                RelOperator::Aggregate(Aggregate {
+                    group_items: p.group_items.clone(),
+                    aggregate_functions: used,
+                    from_distinct: p.from_distinct,
+                }),
+                keep_required_columns(expr.child(0)?, required)?,
+            ))
+        }
+        RelOperator::Sort(p) => {
+            p.items.iter().for_each(|s| {
+                required.insert(s.index);
+            });
+            Ok(SExpr::create_unary(
+                RelOperator::Sort(p.clone()),
+                keep_required_columns(expr.child(0)?, required)?,
+            ))
+        }
+        RelOperator::Limit(p) => Ok(SExpr::create_unary(
+            RelOperator::Limit(p.clone()),
+            keep_required_columns(expr.child(0)?, required)?,
+        )),
+        RelOperator::CrossApply(p) => {
+            for c in &p.correlated_columns {
+                required.insert(*c);
+            }
+            Ok(SExpr::create_binary(
+                RelOperator::CrossApply(p.clone()),
+                keep_required_columns(expr.child(0)?, required)?,
+                prune_columns(expr.child(1)?)?,
```

Comment:
I think you may need to use `keep_required_columns(expr.child(1)?, required)?` for subquery child here.

Reply:
> Sorry for the late reply. It's been a little busy. I think so! But I do not understand `CrossApply` well. Any suggested readings about `CrossApply`?

The idea basically comes from [this paper](https://www.comp.nus.edu.sg/~cs5226/papers/subqueries-sigmod01.pdf), you can take a look if you are interested in it.

The operator performs like a `CrossJoin`, but it will extract the outer refs from its input(i.e. `expr.child(0)`) and apply it for the subquery.

Reply:
Great! I will look at it. Thank you very much.